src/Reflex/Class.hs

```diff
@@ -299,6 +302,32 @@ class ( MonadHold t (PushM t)
   mergeIntIncremental :: Incremental t (PatchIntMap (Event t a)) -> Event t (IntMap a)
   fanInt :: Event t (IntMap a) -> EventSelectorInt t a
 
+-- | Class representing things that can be switched when the provided 'Event' occurs.
+class Reflex t => SwitchHold t a | a -> t where
+  -- | Switches to the new value whenever it is received (via firing of an 'Event').
+  switchHold :: MonadHold t m => a -> Event t a -> m a
+
+-- | Because the simultaneous firing case is irrelevant, this function imposes
+-- laxer "timing requirements" on the overall circuit, avoiding many potential
+-- cyclic dependency / metastability failures. It's also more performant. Use
+-- this rather than 'switchHoldPromptly' and 'switchHoldPromptOnly' unless you
+-- are absolutely sure you need to act on the new event in the coincidental
+-- case.
+instance Reflex t => SwitchHold t (Event t a) where
+  switchHold ea0 eea = switch <$> hold ea0 eea
+
+-- | Switches to the new 'Dynamic' immediately (i.e., promptly) on firing
+instance Reflex t => SwitchHold t (Dynamic t a) where
+  switchHold a e = join <$> holdDyn a e
```

Comment:
I thought join for dynamics was prompt?

Reply:
Something with regular `hold` and `unsafeBuildDynamic` probably has the correct semantics and is the most efficient.

Reply:
Please don't use `unsafeBuildDynamic` unless you can also prove that the result obeys the Dynamic law, i.e. that the Behavior changes when, and only when, the Event fires.

Reply:
Of course! So what I concretely came up with is:
```haskell
import Control.Monad
import Reflex

switchHoldDynamic
  :: (Reflex t, MonadHold t m)
  => Dynamic t a
  -> Event t (Dynamic t a)
  -> m (Dynamic t a)
switchHoldDynamic d ed = switchDynamic <$> hold d ed

-- `switchDyn` and this can duke it out re the naming convention.
switchDynamic
  :: Reflex t
  => Behavior t (Dynamic t a)
  -> Dynamic t a
switchDynamic bd = unsafeBuildDynamic (sample b) e
  where
    b = join $ current <$> bd
    e = switch $ updated <$> bd
```
This leads me to think:
1. Every "unprompt" `switchHold` begins with a plain `hold`. Throwing away the outer event thereafter and just keeping the behavior is crucial to avoiding any promptness / strictness.
2.  If all the interesting logic is in the switch part, perhaps we also want to make `Switch` a class, and today's `switch` rename to `switchE` or something.
3. Single `join` is "`switchB`", `switchDynamic` is actually `unsafeBehaviorEventPairToDyn . bimap switch switch . dynToBehaviorEventPair`.

I won't claim to prove this correct, I think per the above the single `Behavior t (Dynamic t a)` means promptness is impossible, and switching "preserves syncronization" so that switching the underling behavior and event of the inner dynamic maintains the invariants. Put another way, if switching didn't preserve that, I would question the value of `Switch` or `SwitchHold` classes.

Reply:
`switchDynamic` produces an invalid `Dynamic`.

```haskell
-- | If the input event ever fires, the output Dynamic will misbehave.
makeBroken :: (Reflex t, MonadHold t m) => Event t () -> m (Dynamic t Int)
makeBroken changeOuter = do
  outer :: Behavior t (Dynamic t Int) <- hold (pure 0) $ pure 1 <$ changeOuter
  d <- switchDynamic outer
  return d
```

In the invocation of `switchDynamic`, `e` will never fire, because both input `Dynamic`s (`pure 0` and `pure 1`) are constant.  As a result, `updated d` will never fire.  Nevertheless, the value of `current d` will change.

Reply:
Sorry for my wrong-headed attempt. This makes me wonder if there is non-prompt `switchHoldDyn`. The behavior goes with the prompt join of dynamics. The event is suitably not strict/prompt, but doesn't seem to really respect the original dynamic.